File: Python/Problems/Problems/problem890.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set
from collections import Counter

from .problem import Problem


class Problem890(Problem):
    def test(self) -> bool:
        words = ["abc", "deq", "mee", "aqq", "dkd", "ccc"]
        pattern = "abb"

        expected = ["mee", "aqq"]

        result = self.findAndReplacePattern(words, pattern)

        print1DArray(result)

        return expected == result

    def findAndReplacePattern(self, words: List[str], pattern: str) -> List[str]:
        result = []
        ep = self.encode(pattern)
        for word in words:
            if (ep == self.encode(word)):
                result.append(word)

        return result

    def encode(self, word: str) -> str:
        num = 0
        result = ''
        map = dict()
        for c in word:
            if c not in map:
                map[c] = chr(num)
                num += 1
            result += map[c]

        return result

File: C++/Problems/Problems/problem_890.cpp
#include "problem_890.h"

#include <unordered_map>

bool problem_890::test() {
    std::vector<std::string> words { "abc", "deq", "mee", "aqq", "dkd", "ccc" };
    std::string pattern = "abb";

    std::vector<std::string> expected { "mee", "aqq" };

    auto result = findAndReplacePattern(words, pattern);

    print1DVector(result);

    return expected == result;
}

std::vector<std::string> problem_890::findAndReplacePattern(std::vector<std::string> &words, std::string pattern) {
    std::vector<std::string> result;
    auto ep = encode(pattern);
    for (auto word : words)
    {
        if (ep == encode(word))
            result.push_back(word);
    }

    return result;
}

std::string problem_890::encode(std::string word) {
    char num = '0';
    std::string result = "";
    std::unordered_map<char, char> map;
    for (auto c : word)
    {
        if (!map.contains(c))
        {
            map[c] = num;
            num++;
        }

        result += map[c];
    }

    return result;
}
